gh-101100: Fix dangling references in test.rst (#114958)
python/cpython: Doc/library/test.rst

script by itself automatically starts running all regression tests in the
:mod:`test` package. It does this by finding all modules in the package whose
name starts with ``test_``, importing them, and executing the function

> :func:`!test_main` if present or loading the tests via

Reviewer: Or just a code literal, like in the next sentence, since it doesn't refer to a specific named function

```suggestion
``test_main`` if present or loading the tests via
```

Author: In this case, it was referring to a function which the user provided, `test_main`. For people reading the raw rst file, I think the `:func:` prefix provides a small amount of information. It's too bad that disappears in the HTML.

.. data:: Py_DEBUG

>    True if Python is built with the :c:macro:`!Py_DEBUG` macro defined: if

Reviewer 2: It looks like we refer to this macro from quite a few places in the docs:

```
c-api/init_config.rst:      Needs a :ref:`debug build of Python <debug-build>` (the ``Py_DEBUG`` macro
c-api/init_config.rst:      Needs a :ref:`debug build of Python <debug-build>` (the ``Py_DEBUG`` macro
c-api/intro.rst:   If Python is :ref:`built in debug mode <debug-build>` (if the ``Py_DEBUG``
c-api/intro.rst:Compiling the interpreter with the :c:macro:`Py_DEBUG` macro defined produces
c-api/intro.rst::c:macro:`Py_DEBUG` is enabled in the Unix build by adding
c-api/intro.rst:not-Python-specific :c:macro:`_DEBUG` macro.  When :c:macro:`Py_DEBUG` is enabled
library/test.rst:.. data:: Py_DEBUG
library/test.rst:   True if Python is built with the :c:macro:`Py_DEBUG` macro defined: if
using/configure.rst:* Define ``Py_DEBUG`` and ``Py_REF_DEBUG`` macros.
using/configure.rst:* Add runtime checks: code surrounded by ``#ifdef Py_DEBUG`` and ``#endif``.
using/configure.rst:   ``Py_DEBUG`` macro no longer implies the ``Py_TRACE_REFS`` macro (see the
using/configure.rst:   :ref:`Build Python in debug mode <debug-build>`: define the ``Py_DEBUG``
whatsnew/3.10.rst:  <debug-build>` (if the ``Py_DEBUG`` macro is defined). In the limited C API,
whatsnew/3.8.rst:``Py_DEBUG`` macro no longer implies the ``Py_TRACE_REFS`` macro, which
```

Maybe a better way of solving this one would be to add documentation for it, that we can link to from all these places?

Author: `Doc/using/configure.rst` seems to be the best place for this. In fact, there is this:

```
.. _debug-build:

Python Debug Build
------------------

A debug build is Python built with the :option:`--with-pydebug` configure
option.

Effects of a debug build:

...

Debug options
-------------

.. option:: --with-pydebug

   :ref:`Build Python in debug mode <debug-build>`: define the ``Py_DEBUG``
   macro (disabled by default).
```

Given the fact that `Py_DEBUG` is just the marker of a debug build, I'd suggest linking to the `_debug-build` anchor.

.. data:: Py_DEBUG

   True if Python was built with the :c:macro:`!Py_DEBUG` macro defined: if

>    Python is :ref:`built in debug mode <debug-build>`

Reviewer: What's the reason for disabling reference resolution here, sorry? Does it not work, for some reason? A `c:macro:` is defined for `Py_DEBUG`, and the macro definition and surrounding text describes what it means for Python to be built in debug mode, so seems a useful reference here.

Also, this wording change seems a bit incongruous to my eye and the breaks don't either minimize diffs or follow the semantic structure. Therefore, I'd suggest either simplifying the structure and wording with more logical breaks:

```suggestion
   True if Python was built with the :c:macro:`Py_DEBUG` macro defined:
   i.e. it was :ref:`built in debug mode <debug-build>`
```

or minimizing diffs to just the essential fixes

```suggestion
   True if Python was built with the :c:macro:`Py_DEBUG` macro
   defined; that is, if
   Python was :ref:`built in debug mode <debug-build>`
```

Author: Probably just confusion on my part, though I think in discussing changes to `test.rst` we might have gone back and forth a bit. Note that the meaning and effects of debug builds are scattered in multiple locations. In my mind defining the `Py_DEBUG` macro is just the marker for C programmers for what you really care about, creating a debug build. That's discussed in to sections in `using/configure.rst` and `c-api/intro.rst`.

.. data:: MISSING_C_DOCSTRINGS

   Set to ``True`` if Python is built without docstrings (the

>    :c:macro:`!WITH_DOC_STRINGS` macro is not defined).

Reviewer: This appears to be a valid, defined and presumably public macro, it just isn't properly documented yet; therefore, absent extraordinary circumstances we shouldn't silence a valid and useful warning like this. Rather, we should either document it, leave the warning as a reminder to ensure it gets documented in a future pass, or if necessary add it to an appropriate section of the `conf.py` list specifically for public names that we haven't formally documented yet so we can keep track of it in one place.

```suggestion
   :c:macro:`WITH_DOC_STRINGS` macro is not defined).
```

Author: Isn't the reference to the `--without-doc-strings` configure option sufficient? In that section it even states:

```
 Don't define the ``WITH_DOC_STRINGS`` macro.
```

Macros which are or are not defined are nothing magic. They are just the long-term memory of something else the person running `configure` intended. If the details of that meaning exist where the configure option is defined, why duplicate things?

.. function:: with_pymalloc()

>    Return :const:`!_testcapi.WITH_PYMALLOC`.

Reviewer: This is a constant in an underscored name, test module, so it's not required to document it, but given the entire description of this function is "return <constant>", which doesn't seem terribly useful if said constant is entirely undocumented, it seems worth at least briefly touching on what that actually means here (which the warning is a canary for).

Author: I'm guessing the reference must be suppressed in some other way, because no hyperlink shows up in the generated HTML (and no warning is emitted by Sphinx). You'll have to take this up with someone above my pay grade.

.. function:: skip_if_broken_multiprocessing_synchronize()

>    Skip tests if the :mod:`!multiprocessing.synchronize` module is missing, if

Reviewer: This is an purposefully undocumented module, I'm guessing?

Author: I couldn't begin to guess why it's undocumented. Maybe it's meant to be semantically identical to the relevant `threading` module synchronization primitives?

```
__all__ = [
    'Lock', 'RLock', 'Semaphore', 'BoundedSemaphore', 'Condition', 'Event'
    ]
```

I see this comment preceding an import which it suggests might fail:

```
# Try to import the mp.synchronize module cleanly, if it fails
# raise ImportError for platforms lacking a working sem_open implementation.
# See issue 3770
```

There is also a note about this possibility in the `multiprocessing` module documentation.

   buildbot environment.  This method raises an exception if the
   ``sock.family`` is :const:`~socket.AF_INET` and ``sock.type`` is
   :const:`~socket.SOCK_STREAM`, and the socket has

>    :const:`!SO_REUSEADDR` or :const:`!SO_REUSEPORT` set on it.

Reviewer 2: The better way to fix these two warnings would be to add documentation for these constants in the `socket` module docs, if that documentation doesn't already exist. Then we'll also be able to reference these constants in the docs for other modules, as will users of intersphinx outside of CPython

Author: Perhaps, but ISTR the `socket` module doc explicitly mentions a whole raft of constants as coming from the underlying C socket(2) implementation. That is presumably to keep from duplicating underlying canonical documentation and to avoid introducing mistakes (think platform dependencies).

I understand it would be great to have the Python library docs completely self-contained, but I don't think that's the best way to go in some cases.

Reviewer 2: Fair point, there are probably too many in the socket module for us to list them all out explicitly!

Reviewer: In that case, we should explicitly list them in `conf.py` as silenced with a comment explaining why, to avoid introducing noise here, making it easy to change the decision in the future, and ensuring we have a consistent central record of it.

Author: Given that many of the possible constants are only given in the `socket` module doc as wildcards (e.g., `SO_*`) I don't think that will work unless you only want to list a known (small) subset.

Author: This stuff is really at the fringe between low-level Unix stuff and the relevant Python wrappers. According to Wikipedia, the first edition of Stevens' _Unix Network Programming_ (the network programming bible when I was a wee lad) is 768 pages long. It's not at all clear to me that there's much to be gained by documenting a few socket options. It's not really going to flatten the socket programming learning curve much. Programmers will have to delve into the C socket documentation to do anything serious anyway.

Sorry about the soapbox.

   In this case all warnings are caught and no errors are raised.

   On entry to the context manager, a :class:`WarningsRecorder` instance is
   returned. The underlying warnings list from
   :func:`~warnings.catch_warnings` is available via the recorder object's

>    :attr:`warnings` attribute.  As a convenience, the attributes of the object

Reviewer: This existing reference resolves but entirely incorrectly; it implies `warnings` is an "attribute" of the `test.support.warnings_helper` module, and furthermore actually falls back to link to the `warnings` stdlib module instead, which is entirely not what is intended. It should presumably be moved to where it belongs, under the class below, instead, to both avoid the misleading reference and properly document it.

Author: This is one place where it would be real nice if Sphinx applied some simple semantics to the refs. Something which is an attribute clearly can't be a module. I've qualified the reference. As with the `reset` reference, it is now broken.

```
.../Doc/library/test.rst:1719: WARNING: py:attr reference target not found: WarningsRecorder.warnings
```

Both the `reset` and `warnings` items are fully qualified, so even though the links don't work, the full names are displayed.

   then any of the attributes that would otherwise be expected on an object
   representing a warning will return ``None``.

   The recorder object also has a :meth:`!reset` method, which clears the

>    warnings list.

Reviewer: This is a valid, public method of `WarningsRecorder`, a class publicly documented just below, that is missing its documentation. Therefore, this warning shouldn't be silenced, as again it indicates a legitimite gap in the docs. The simplest solution seems to move this line to a `.. method` under the `WarningsRecorder` class to properly document the method there, instead of overloading the already long and complex `check_warnings` context description with details of the object it returns.

Author: I changed the text to `:meth:``~WarningsRecorder.reset`` ` (how in the heck do I make the raw ReST display correctly in Markdown???) which is fine as far as that goes, but there is no documentation of either the class or the reset method. I'll leave it as-is. There will be a nitpick warning:

```
.../Doc/library/test.rst:1728: WARNING: py:meth reference target not found: WarningsRecorder.reset
```

Reviewer 3: > (how in the heck do I make the raw ReST display correctly in Markdown???)

Use the fact that GitHub supports HTML tags and escape backquotes:

```html
<code>:meth:\`~WarningsRecorder.reset\`</code>
```

produces <code>:meth:\`~WarningsRecorder.reset\`</code>.
